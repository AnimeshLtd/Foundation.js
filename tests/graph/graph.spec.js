const { Graph } = require("../../source")

let graph;

describe("Graph", function() {
  beforeEach(() => {
    graph = new Graph()
  })

  it("should initialise",                       testInit)
  it("should add vertex",                       testAddVertex)
  it("should remove vertex",                    testRemoveVertex)
  it("should add edge for a directed graph",    testAddEdgeDirected)
  it("should add edge for an undirected graph", testAddEdgeUndirected)
  it("should remove edge for directed graph",   testRemoveEdgeDirected)
  it("should remove edge for undirected graph", testRemoveEdgeUndirected)
  it("should search depth-first",               testDepthFirstSearch)
  it("should search breadth-first",             testBreadthFirstSearch)
})

function testInit() {
  expect(graph._isDirected).toBe(true)
  expect(graph.count).toBe(0)
}

function testAddVertex() {
  graph.addVertex("Mongolia")
  graph.addVertex("India")
  graph.addVertex("Kenya")
  expect(graph.count).toBe(3)
}

function testRemoveVertex() {
  graph.addVertex("Mongolia")
  graph.addVertex("India")
  graph.addVertex("Kenya")
  expect(graph.count).toBe(3)

  const removed = graph.removeVertex("Kenya")
  expect(graph.count).toBe(2)
  expect(removed.data.key).toBe("Kenya")
}

function testAddEdgeDirected() {
  const { source, destination } = graph.addEdge("India", "Mongolia")
  expect(source.isNeighbour(destination)).toBe(true)
  expect(destination.isNeighbour(source)).toBe(false)
  expect(graph.areNeighbours("India", "Mongolia")).toBe(true)
}

function testAddEdgeUndirected() {
  const undirected = new Graph(false)
  const { source, destination } = undirected.addEdge("India", "Mongolia")
  expect(source.isNeighbour(destination)).toBe(true)
  expect(destination.isNeighbour(source)).toBe(true)
  expect(undirected.areNeighbours("India", "Mongolia")).toBe(true)
}

function testRemoveEdgeDirected() {
  const { source, destination } = graph.addEdge("India", "Mongolia")
  expect(source.isNeighbour(destination)).toBe(true)
  expect(destination.isNeighbour(source)).toBe(false)
  expect(graph.areNeighbours("India", "Mongolia")).toBe(true)

  expect(graph.removeEdge("India", "Mongolia")).toBe(true)
  expect(source.isNeighbour(destination)).toBe(false)
  expect(destination.isNeighbour(source)).toBe(false)
  expect(graph.areNeighbours("India", "Mongolia")).toBe(false)
}

function testRemoveEdgeUndirected() {
  const undirected = new Graph(false)
  const { source, destination } = undirected.addEdge("India", "Mongolia")
  expect(source.isNeighbour(destination)).toBe(true)
  expect(destination.isNeighbour(source)).toBe(true)
  expect(undirected.areNeighbours("India", "Mongolia")).toBe(true)

  expect(undirected.removeEdge("India", "Mongolia")).toBe(true)
  expect(source.isNeighbour(destination)).toBe(false)
  expect(destination.isNeighbour(source)).toBe(false)
  expect(undirected.areNeighbours("India", "Mongolia")).toBe(false)
}

function testDepthFirstSearch() {
  graph.addVertex("Mongolia")
  graph.addVertex("India")
  graph.addVertex("Kenya")
  graph.addVertex("Zimbabwe")
  graph.addVertex("Austria")
  graph.addVertex("Sweden")
  graph.addEdge("Mongolia", "India")
  graph.addEdge("Mongolia", "Austria")
  graph.addEdge("India", "Sweden")
  graph.addEdge("India", "Austria")
  graph.addEdge("India", "Zimbabwe")
  graph.addEdge("Kenya", "Zimbabwe")
  graph.addEdge("Austria", "India")
  graph.addEdge("Austria", "Sweden")
  graph.addEdge("Sweden", "Mongolia")

  expect(Array.from(graph.searchDepthFirst("India"))).toMatchSnapshot()
}

function testBreadthFirstSearch() {
  graph.addVertex("Mongolia")
  graph.addVertex("India")
  graph.addVertex("Kenya")
  graph.addVertex("Zimbabwe")
  graph.addVertex("Austria")
  graph.addVertex("Sweden")
  graph.addEdge("Mongolia", "India")
  graph.addEdge("Mongolia", "Austria")
  graph.addEdge("India", "Sweden")
  graph.addEdge("India", "Austria")
  graph.addEdge("India", "Zimbabwe")
  graph.addEdge("Kenya", "Zimbabwe")
  graph.addEdge("Austria", "India")
  graph.addEdge("Austria", "Sweden")
  graph.addEdge("Sweden", "Mongolia")

  expect(Array.from(graph.searchBreadthFirst("India"))).toMatchSnapshot()
}
