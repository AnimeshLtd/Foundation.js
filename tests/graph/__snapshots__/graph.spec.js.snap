// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph should search breadth-first 1`] = `
Array [
  GraphNode {
    "_data": "India",
    "_neighbours": Set {
      GraphNode {
        "_data": "Sweden",
        "_neighbours": Set {
          GraphNode {
            "_data": "Mongolia",
            "_neighbours": Set {
              [Circular],
              GraphNode {
                "_data": "Austria",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
            },
          },
        },
      },
      GraphNode {
        "_data": "Austria",
        "_neighbours": Set {
          [Circular],
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              GraphNode {
                "_data": "Mongolia",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
            },
          },
        },
      },
      GraphNode {
        "_data": "Zimbabwe",
        "_neighbours": Set {},
      },
    },
  },
  GraphNode {
    "_data": "Sweden",
    "_neighbours": Set {
      GraphNode {
        "_data": "Mongolia",
        "_neighbours": Set {
          GraphNode {
            "_data": "India",
            "_neighbours": Set {
              [Circular],
              GraphNode {
                "_data": "Austria",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
              GraphNode {
                "_data": "Zimbabwe",
                "_neighbours": Set {},
              },
            },
          },
          GraphNode {
            "_data": "Austria",
            "_neighbours": Set {
              GraphNode {
                "_data": "India",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                  GraphNode {
                    "_data": "Zimbabwe",
                    "_neighbours": Set {},
                  },
                },
              },
              [Circular],
            },
          },
        },
      },
    },
  },
  GraphNode {
    "_data": "Austria",
    "_neighbours": Set {
      GraphNode {
        "_data": "India",
        "_neighbours": Set {
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              GraphNode {
                "_data": "Mongolia",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
            },
          },
          [Circular],
          GraphNode {
            "_data": "Zimbabwe",
            "_neighbours": Set {},
          },
        },
      },
      GraphNode {
        "_data": "Sweden",
        "_neighbours": Set {
          GraphNode {
            "_data": "Mongolia",
            "_neighbours": Set {
              GraphNode {
                "_data": "India",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                  GraphNode {
                    "_data": "Zimbabwe",
                    "_neighbours": Set {},
                  },
                },
              },
              [Circular],
            },
          },
        },
      },
    },
  },
  GraphNode {
    "_data": "Zimbabwe",
    "_neighbours": Set {},
  },
  GraphNode {
    "_data": "Mongolia",
    "_neighbours": Set {
      GraphNode {
        "_data": "India",
        "_neighbours": Set {
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              [Circular],
            },
          },
          GraphNode {
            "_data": "Austria",
            "_neighbours": Set {
              [Circular],
              GraphNode {
                "_data": "Sweden",
                "_neighbours": Set {
                  [Circular],
                },
              },
            },
          },
          GraphNode {
            "_data": "Zimbabwe",
            "_neighbours": Set {},
          },
        },
      },
      GraphNode {
        "_data": "Austria",
        "_neighbours": Set {
          GraphNode {
            "_data": "India",
            "_neighbours": Set {
              GraphNode {
                "_data": "Sweden",
                "_neighbours": Set {
                  [Circular],
                },
              },
              [Circular],
              GraphNode {
                "_data": "Zimbabwe",
                "_neighbours": Set {},
              },
            },
          },
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              [Circular],
            },
          },
        },
      },
    },
  },
]
`;

exports[`Graph should search depth-first 1`] = `
Array [
  GraphNode {
    "_data": "India",
    "_neighbours": Set {
      GraphNode {
        "_data": "Sweden",
        "_neighbours": Set {
          GraphNode {
            "_data": "Mongolia",
            "_neighbours": Set {
              [Circular],
              GraphNode {
                "_data": "Austria",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
            },
          },
        },
      },
      GraphNode {
        "_data": "Austria",
        "_neighbours": Set {
          [Circular],
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              GraphNode {
                "_data": "Mongolia",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
            },
          },
        },
      },
      GraphNode {
        "_data": "Zimbabwe",
        "_neighbours": Set {},
      },
    },
  },
  GraphNode {
    "_data": "Zimbabwe",
    "_neighbours": Set {},
  },
  GraphNode {
    "_data": "Austria",
    "_neighbours": Set {
      GraphNode {
        "_data": "India",
        "_neighbours": Set {
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              GraphNode {
                "_data": "Mongolia",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
            },
          },
          [Circular],
          GraphNode {
            "_data": "Zimbabwe",
            "_neighbours": Set {},
          },
        },
      },
      GraphNode {
        "_data": "Sweden",
        "_neighbours": Set {
          GraphNode {
            "_data": "Mongolia",
            "_neighbours": Set {
              GraphNode {
                "_data": "India",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                  GraphNode {
                    "_data": "Zimbabwe",
                    "_neighbours": Set {},
                  },
                },
              },
              [Circular],
            },
          },
        },
      },
    },
  },
  GraphNode {
    "_data": "Sweden",
    "_neighbours": Set {
      GraphNode {
        "_data": "Mongolia",
        "_neighbours": Set {
          GraphNode {
            "_data": "India",
            "_neighbours": Set {
              [Circular],
              GraphNode {
                "_data": "Austria",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                },
              },
              GraphNode {
                "_data": "Zimbabwe",
                "_neighbours": Set {},
              },
            },
          },
          GraphNode {
            "_data": "Austria",
            "_neighbours": Set {
              GraphNode {
                "_data": "India",
                "_neighbours": Set {
                  [Circular],
                  [Circular],
                  GraphNode {
                    "_data": "Zimbabwe",
                    "_neighbours": Set {},
                  },
                },
              },
              [Circular],
            },
          },
        },
      },
    },
  },
  GraphNode {
    "_data": "Mongolia",
    "_neighbours": Set {
      GraphNode {
        "_data": "India",
        "_neighbours": Set {
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              [Circular],
            },
          },
          GraphNode {
            "_data": "Austria",
            "_neighbours": Set {
              [Circular],
              GraphNode {
                "_data": "Sweden",
                "_neighbours": Set {
                  [Circular],
                },
              },
            },
          },
          GraphNode {
            "_data": "Zimbabwe",
            "_neighbours": Set {},
          },
        },
      },
      GraphNode {
        "_data": "Austria",
        "_neighbours": Set {
          GraphNode {
            "_data": "India",
            "_neighbours": Set {
              GraphNode {
                "_data": "Sweden",
                "_neighbours": Set {
                  [Circular],
                },
              },
              [Circular],
              GraphNode {
                "_data": "Zimbabwe",
                "_neighbours": Set {},
              },
            },
          },
          GraphNode {
            "_data": "Sweden",
            "_neighbours": Set {
              [Circular],
            },
          },
        },
      },
    },
  },
]
`;
